### PR TITLE
Pal 650 resource service redis persistence

### PR DESCRIPTION
**Pull request by dev930018**

### CODE REVIEW

File: resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/domain/CompletenessEntity.java [hunk 1]
```diff
@@ -1,111 +1,111 @@
...
+    protected String entityId;
+
+    public CompletenessEntity() {
+    }
+
+    @PersistenceConstructor
+    public CompletenessEntity(final EntityType entityType, final String entityId) {
+        this.entityType = entityType;
+        this.entityId = entityId;
+        this.id = hashCode();
```
Comment (dev930018) [inline resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/domain/CompletenessEntity.java]: Could see this being `entityType.toString() + "::" + entityId`

File: resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/domain/ResourceEntity.java [hunk 2]
```diff
@@ -39,11 +44,15 @@
                 @Index(name = "resource_id", columnList = "resource_id", unique = true),
                 @Index(name = "parent_id", columnList = "parent_id"),
         })
-public class ResourceEntity {
+@RedisHash("ResourceEntity")
+public class ResourceEntity implements Serializable {
     @Id
+    @org.springframework.data.annotation.Id
```
Comment (pd104923) [inline resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/domain/ResourceEntity.java:57]: What's the need to have both of these Id annotations?

Comment (dev930018) [reply resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/domain/ResourceEntity.java:57]: Redis requires a separate set of spring annotations to h2 and javax annotations

File: resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/domain/CompletenessEntity.java (hunk 1 above)
Comment (pd104923) [reply resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/domain/CompletenessEntity.java]: I'm going to change this as SQ doesn't like using `hash()` in the constructor.

Comment (pd104923) [reply resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/domain/CompletenessEntity.java]: The implemented change caused a test issue. Left it for now.

File: resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/service/StreamingResourceServiceProxy.java [hunk 3]
```diff
@@ -1,195 +1,196 @@
...
+import uk.gov.gchq.palisade.util.ResourceBuilder;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+
```
Comment (developer01189998819991197253) [inline resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/service/StreamingResourceServiceProxy.java:42]: javadoc required

File: resource-service/src/test/java/uk/gov/gchq/palisade/service/resource/repository/JpaPersistenceLayerTest.java [hunk 4]
```diff
@@ -1,117 +1,117 @@
-/*
```
Comment (developer01189998819991197253) [inline resource-service/src/test/java/uk/gov/gchq/palisade/service/resource/repository/JpaPersistenceLayerTest.java:1]: There are no changes made to this file

File: resource-service/src/test/java/uk/gov/gchq/palisade/service/resource/service/SimpleResourceServiceTest.java [hunk 5]
```diff
@@ -1,88 +1,88 @@
-/*
```
Comment (developer01189998819991197253) [inline resource-service/src/test/java/uk/gov/gchq/palisade/service/resource/service/SimpleResourceServiceTest.java:1]: There are no changes made to this file

**Review by developer01189998819991197253 — CHANGES_REQUESTED** [review]
A few files have no changes made to them so should be removed from the changelist

File: resource-service/src/test/java/uk/gov/gchq/palisade/service/resource/service/SimpleResourceServiceTest.java (hunk 5 above)
Comment (dev930018) [reply resource-service/src/test/java/uk/gov/gchq/palisade/service/resource/service/SimpleResourceServiceTest.java:1]: Line endings, \r\n to \n. Only apply if there were other changes to the file?

File: resource-service/src/test/java/uk/gov/gchq/palisade/service/resource/repository/JpaPersistenceLayerTest.java (hunk 4 above)
Comment (dev930018) [reply resource-service/src/test/java/uk/gov/gchq/palisade/service/resource/repository/JpaPersistenceLayerTest.java:1]: Line endings, \r\n to \n. Only apply if there were other changes to the file?

**Review by developer01189998819991197253 — DISMISSED** [review]

**Review by ac74475 — DISMISSED** [review]

**Review by dbz120896 — DISMISSED** [review]

**Review by dev-958 — COMMENTED** [review]
I notice that you have used Spring Boot Starter Redis, which autoconfigures Redis for you, it may be sufficient for this use case but it may have been better to use the more feature-rich Spring Data Redis (RedisTemplate), which is what I was expecting here. Some explanations of the entity structures in the README would be useful as they are used in controlling the lifecycle of the data. I mean create a README for just this module.

**Comment by dev930018** [comment]
https://github.com/gchq/Palisade-services/pull/169#pullrequestreview-483355162

In this case I was happy with spring-boot-starter-redis and don't think any more control is needed (yet). Depends how much future-proofing you want here.

README sounds like a good call at this point, will address.

File: resource-service/README.md [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+|  resource: JSON         | --------------------------- ---------------------------
+---------------------------
+```
+(`*` marks an indexable field)
+
+### Motivation
+To ease load on an external resource-service provider, a cache-like storage mechanism is used to return data for previously-queried resources.
+The code for this can be found in the [StreamingResourceServiceProxy](src/main/java/uk/gov/gchq/palisade/service/resource/service/StreamingResourceServiceProxy.java) and [JpaPersistenceLayer](src/main/java/uk/gov/gchq/palisade/service/resource/repository/JpaPersistenceLayer.java).
+
+Due to the tree-structure of filesystems, naive caching of requests-responses results in a large amount of data while not providing any "smart" methods to the caching mechanism.
```
Comment (developer01189998819991197253) [inline resource-service/README.md]:
```suggestion
Due to the tree-structure of filesystems, naïve caching of requests-responses results in a large amount of data while not providing any "smart" methods to the caching mechanism.
```

**Review by developer01189998819991197253 — CHANGES_REQUESTED** [review]
Spelling mistake in readme but otherwise fine

File: resource-service/README.md (hunk 6 above)
Comment (dev930018) [reply resource-service/README.md]: I meant naive, i can ï my i if you like for clarity

**Review by developer01189998819991197253 — APPROVED** [review]

**Review by ac74475 — APPROVED** [review]

**Review by dbz120896 — APPROVED** [review]

**Pull request merged**